test/middleware/csp.test.js

```diff
@@ -1,6 +1,147 @@
+  underTest.set('trust proxy', 1)
+  underTest.use('/test', cspMiddlewareStack)
+
+  beforeEach(() => {
+    loggingSpy.resetHistory()
+    sentrySpy.resetHistory()
+  })
+
+  const validPayload = {
+    'csp-report': {
+      'blocked-uri': 'https://www.example.com',
+      'violated-directive': 'connect-src'
+    }
+  }
+
+  it('should return 204 and send to sentry if a valid csp report is present', async () => {
+    await request(underTest)
+      .post('/test')
+      .set('user-agent', 'supertest')
+      .send(validPayload)
+      .expect(204)
+
+    expect(sentrySpy.calledOnce).to.be.true
+    expect(sentrySpy.calledWith({
+      message: 'Blocked connect-src from https://www.example.com',
+      level: 'warning',
+      extra: {
+        'cspReport': validPayload['csp-report'],
+        'userAgent': 'supertest'
+      }
+    })).to.be.true
+  })
+
+  it('should return 204 and not send to sentry if a valid csp report is present but the blocked-uri is ignored', async () => {
+    await request(underTest)
+      .post('/test')
+      .set('user-agent', 'supertest')
+      .send({
+        'csp-report': {
+          'blocked-uri': 'https://www.banned.com',
+          'violated-directive': 'connect-src'
+        }
+      })
+      .expect(204)
+
+    expect(sentrySpy.notCalled).to.be.true
+  })
+
+  it('should return 400 if content-type is unexpected', async () => {
+    await request(underTest)
+      .post('/test')
+      .set('content-type', 'application/x-www-form-urlencoded')
+      .send(validPayload)
+      .expect(400)
+  })
+
+  it('should return 400 if payload is too large', async () => {
+
+    const largePayload = {
+      ...validPayload,
+      'large_value': 'some text here x1000'.repeat(1000)
+    }
+
+    await request(underTest)
+      .post('/test')
+      .send(largePayload)
+      .expect(400)
+
+    expect(loggingSpy.calledOnce).to.be.true
+    expect(loggingSpy.calledWith('CSP violation request payload exceeds maximum size')).to.be.true
+  })
+
+  it('should return 400 if request is not JSON', async () => {
+    await request(underTest)
+      .post('/test')
+      .set('content-type', 'application/json')
+      .send('notJSON')
+      .expect(400)
+
+    expect(loggingSpy.calledOnce).to.be.true
+    expect(loggingSpy.calledWith('CSP violation request payload could not be parsed')).to.be.true
+  })
+
+  it('should return 400 if json does not included expected values', async () => {
+    await request(underTest)
+      .post('/test')
+      .send({ key: 'value' })
+      .expect(400)
+  })
+
+  it('should return 429 if too many requests are made', async () => {
+    for (let i = 0; i < 10; i++) {
+      await request(underTest)
+        .post('/test')
+        .set('x-forwarded-for', '9.9.9.9')
+        .send(validPayload)
+        .expect(204)
+    }
+
+    await request(underTest)
+      .post('/test')
+      .set('x-forwarded-for', '9.9.9.9')
+      .send(validPayload)
+      .expect(429)
```

Comment:
As discussed, let's have a test for the level 2 `trust proxy`  (given that's what we'll be using).

Reply:
yep, the tests did not adequately validate this behaviour, have amended